### PR TITLE
WD-1053 - Create general slice

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Move general actions, reducers, types and selectors to a slice.
- Add tests for the above.

## QA

- Pull code
- Run `dotrun clean && dotrun serve`
- Open http://0.0.0.0:8036/
- Run the app and check that you can log in and list models etc.

## Details

https://warthogs.atlassian.net/browse/WD-1053

### CODE REVIEW

File: src/app/actions.js [hunk 1]
```diff
@@ -1,103 +1,20 @@
...
-
+import { actions as generalActions } from "store/general";
 import {
   getConfig,
   getControllerConnections,
+  getPingerIntervalIds,
   getUserPass,
   getWSControllerURL,
-} from "./selectors";
-import { actionsList } from "./action-types";
```
Comment (huwshimi) [inline src/app/actions.js:16]: These actions are now created in the slice.

File: src/app/general.js [hunk 2]
```diff
@@ -1,47 +0,0 @@
-import immerProduce from "immer";
-import cloneDeep from "clone-deep";
-
-import { actionsList } from "./action-types";
-
-function generalReducer(state = {}, action) {
```
Comment (huwshimi) [inline src/app/general.js:6]: These reducers are now handled in the slice.

File: src/app/selectors.js [hunk 3]
```diff
@@ -31,45 +31,6 @@ export const getModelData = (state) => {
 */
 export const getControllerData = (state) => state?.juju?.controllers;
 
-/**
```
Comment (huwshimi) [inline src/app/selectors.js:34]: These selectors have been extracted to the general slice file and updated to use `createSelector`.

File: src/components/LogIn/LogIn.tsx [hunk 4]
```diff
@@ -36,7 +37,7 @@ export default function LogIn({ children }: PropsWithChildren<ReactNode>) {
   // Loop through all of the available controller connections to see
   // if we're logged in.
   const userIsLoggedIn = wsControllerURLs.some((wsControllerURL) =>
-    isLoggedIn(wsControllerURL, store.getState())
+    isLoggedIn(store.getState(), wsControllerURL)
```
Comment (huwshimi) [inline src/components/LogIn/LogIn.tsx:40]: `createSelector` requires the state to be first.

File: src/store/general/actions.test.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import { actions } from "./slice";
+
+describe("selectors", () => {
```
Comment (huwshimi) [inline src/store/general/actions.test.ts]: These tests are all new.

File: src/store/general/reducers.test.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
+import { actions, reducer } from "./slice";
+
+const defaultState = {
```
Comment (huwshimi) [inline src/store/general/reducers.test.ts:3]: These tests will be updated to use factory functions in a followup (https://warthogs.atlassian.net/browse/WD-822).

File: src/store/general/reducers.test.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  appVersion: null,
+  config: null,
+  controllerConnections: null,
+  credentials: null,
+  loginError: null,
+  pingerIntervalIds: null,
+  visitURL: null,
+};
+
+describe("selectors", () => {
```
Comment (huwshimi) [inline src/store/general/reducers.test.ts]: These tests are new.

File: src/store/general/selectors.test.ts [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+  appVersion: null,
+  config: null,
+  controllerConnections: null,
+  credentials: null,
+  loginError: null,
+  pingerIntervalIds: null,
+  visitURL: null,
+};
+
+describe("selectors", () => {
```
Comment (huwshimi) [inline src/store/general/selectors.test.ts:27]: These tests are new

File: src/types.ts [hunk 9]
```diff
@@ -8,30 +8,6 @@ export type JujuState = {
   modelWatcherData: ModelWatcherData;
 };
 
-export type Config = {
```
Comment (huwshimi) [inline src/types.ts:11]: These types are now in the slice directory.

**Review by goulinkh — APPROVED** [review]

**Pull request merged**